File: js/process_transactions.js
const processTransactions = (transactions) => {
  const result = {};
  const seenIds = new Set();
  for (const txn of transactions) {
    // skip transaction if we have seen it already
    if (seenIds.has(txn.id)) continue;

    const { userId, amount, type } = txn;

    // initial balance in the result if it doesn't exist already
    if (!result[userId]) {
      result[userId] = 0;
    }

    // adjust balance
    type === 'credit' ? result[userId] += amount : result[userId] -= amount;

    // add id to seenIds set
    seenIds.add(txn.id);
  }

  return result
};

const reconcileTxn = (txn1, txn2) => {
  const { timestamp: time1, amount: amount1, type: type1 } = txn1;
  const { timestamp: time2, amount: amount2, type: type2 } = txn2;
  const totalVal1 = type1 === 'credit' ? amount1 : -amount1;
  const totalVal2 = type2 === 'credit' ? amount2 : -amount2;

  return time1 > time2 ? totalVal1 - totalVal2 : totalVal2 - totalVal1;
};

const txns = [
  { id: 1, userId: "A", amount: 100, type: "credit", timestamp: 1712520000000 },
  { id: 2, userId: "B", amount: 50, type: "credit", timestamp: 1712520060000 },
  { id: 1, userId: "A", amount: 120, type: "credit", timestamp: 1712520120000 }, // Newer duplicate ID 1
  { id: 3, userId: "A", amount: 20, type: "debit", timestamp: 1712520180000 },
  { id: 4, userId: "C", amount: 1000, type: "credit", timestamp: 1712520240000 },
  { id: 5, userId: "B", amount: 25, type: "debit", timestamp: 1712520300000 },
  { id: 6, userId: "A", amount: 50, type: "credit", timestamp: 1712520360000 },
  { id: 4, userId: "C", amount: 1000, type: "credit", timestamp: 1712520000000 }, // Older duplicate ID 4
  { id: 7, userId: "C", amount: 150, type: "debit", timestamp: 1712520480000 },
  { id: 8, userId: "D", amount: 300, type: "credit", timestamp: 1712520540000 }
];

const processTransactionsWithTimestamps = transactions => {
  const result = {};
  const seenIds = new Map();

  for (const txn of transactions) {
    const { id, userId, amount, type } = txn;

    // initial balance in the result if it doesn't exist already
    if (!result[userId]) {
      result[userId] = 0;
    }

    // adjust balance
    if (seenIds.has(id)) {
      const diff = reconcileTxn(txn, seenIds.get(id));
      result[userId] += diff;
    } else {
      type === 'credit' ? result[userId] += amount : result[userId] -= amount;
    }

    // add id to seeIds or override existing id
    seenIds.set(id, txn);
  };

  return result;
};

console.log(processTransactionsWithTimestamps(txns));

File: python/process_transactions.py
from collections import defaultdict

txns2 = [
    {"id": 1, "user_id": "A", "amount": 100,
        "type": "credit", "timestamp": 1712520000000},
    {"id": 2, "user_id": "B", "amount": 50,
        "type": "credit", "timestamp": 1712520060000},
    {"id": 1, "user_id": "A", "amount": 120, "type": "credit",
        "timestamp": 1712520120000},  # Newer duplicate ID 1
    {"id": 3, "user_id": "A", "amount": 20,
        "type": "debit", "timestamp": 1712520180000},
    {"id": 4, "user_id": "C", "amount": 1000,
        "type": "credit", "timestamp": 1712520240000},
    {"id": 5, "user_id": "B", "amount": 25,
        "type": "debit", "timestamp": 1712520300000},
    {"id": 6, "user_id": "A", "amount": 50,
        "type": "credit", "timestamp": 1712520360000},
    {"id": 4, "user_id": "C", "amount": 1000, "type": "credit",
        "timestamp": 1712520000000},  # Older duplicate ID 4
    {"id": 7, "user_id": "C", "amount": 150,
        "type": "debit", "timestamp": 1712520480000},
    {"id": 8, "user_id": "D", "amount": 300,
        "type": "credit", "timestamp": 1712520540000}
]


def reconcile_txn(txn1, txn2):
    time1, amount1, type1 = txn1['timestamp'], txn1['amount'], txn1['type']
    time2, amount2, type2 = txn2['timestamp'], txn2['amount'], txn2['type']

    total_val1 = amount1 if type1 == 'credit' else -amount1
    total_val2 = amount2 if type2 == 'credit' else -amount2

    if time1 > time2:
        return total_val1 - total_val2
    else:
        return total_val2 - total_val1


def process_transactions(txns):
    result = defaultdict(int)
    seen_ids = {}

    for txn in txns:
        id, user_id = txn["id"], txn["user_id"]

        if id in seen_ids:
            diff = reconcile_txn(txn, seen_ids[id])
            result[user_id] += diff
        else:
            amount, txn_type = txn["amount"], txn["type"]

            if txn_type == 'credit':
                result[user_id] += amount
            else:
                result[user_id] -= amount

        seen_ids[id] = txn

    return result


print(process_transactions(txns2))
